src/worker.c

```diff
@@ -255,64 +255,99 @@ worker_main(Datum main_arg)
 						curl_easy_getinfo(eh, CURLINFO_PRIVATE, &id);
-						curl_easy_getinfo(eh, CURLINFO_CONTENT_TYPE, &contentType);
-
-						elog(DEBUG2, "GET of %ld returned http status code %d\n", id, http_status_code);
-
-						cdata = hash_search(curlDataMap, &id, HASH_FIND, &isPresent);
-
-						argTypes[0] = INT8OID;
-						argValues[0] = Int64GetDatum(id);
 
-						argTypes[1] = INT4OID;
-						argValues[1] = Int32GetDatum(http_status_code);
-
-						argTypes[2] = CSTRINGOID;
-						argValues[2] = CStringGetDatum(cdata->body->data);
-
-						argTypes[3] = JSONBOID;
-						argValues[3] = JsonbPGetDatum(JsonbValueToJsonb(pushJsonbValue(&cdata->headers, WJB_END_OBJECT, NULL)));
-
-						argTypes[4] = CSTRINGOID;
-						argValues[4] = CStringGetDatum(contentType);
-
-						argTypes[5] = BOOLOID;
-						argValues[5] = BoolGetDatum(timedOut);
-
-						if (SPI_execute_with_args(insert_query.data, argCount, argTypes, argValues, NULL,
-										false, 1) != SPI_OK_INSERT)
-						{
-							elog(ERROR, "SPI_exec failed: %s", insert_query.data);
+						if (return_code != CURLE_OK) {
+							int argCount = 2;
+							Oid argTypes[2];
+							Datum argValues[2];
+							const char *error_msg = curl_easy_strerror(return_code);
+
+							argTypes[0] = CSTRINGOID;
+							argValues[0] = CStringGetDatum(error_msg);
+
+							argTypes[1] = INT8OID;
+							argValues[1] = Int64GetDatum(id);
+
+							elog(DEBUG1, "%s\n", error_msg);
+
+							if (SPI_execute_with_args(update_query.data, argCount, argTypes, argValues, NULL, false, 1) != SPI_OK_UPDATE)
+							{
+								elog(ERROR, "SPI_exec failed: %s", update_query.data);
+							}
+						} else {
+							int argCount = 6;
+							Oid argTypes[6];
+							Datum argValues[6];
+							char nulls[6];
+							CurlData *cdata = NULL;
+							char *contentType = NULL;
+							bool timedOut = false;
+
+							curl_easy_getinfo(eh, CURLINFO_RESPONSE_CODE, &http_status_code);
+							curl_easy_getinfo(eh, CURLINFO_CONTENT_TYPE, &contentType);
+							curl_easy_getinfo(eh, CURLINFO_PRIVATE, &id);
+
+							elog(DEBUG1, "GET of %ld returned http status code %d\n", id, http_status_code);
+
+							cdata = hash_search(curlDataMap, &id, HASH_FIND, &isPresent);
+
+							argTypes[0] = INT8OID;
+							argValues[0] = Int64GetDatum(id);
+							nulls[0] = ' ';
+
+							argTypes[1] = INT4OID;
+							argValues[1] = Int32GetDatum(http_status_code);
+							nulls[1] = ' ';
+
+							argTypes[2] = CSTRINGOID;
+							argValues[2] = CStringGetDatum(cdata->body->data);
+							if(cdata->body->data[0] == '\0')
+								nulls[2] = 'n';
+							else
+								nulls[2] = ' ';
+
+							argTypes[3] = JSONBOID;
+							argValues[3] = JsonbPGetDatum(JsonbValueToJsonb(pushJsonbValue(&cdata->headers, WJB_END_OBJECT, NULL)));
+							nulls[3] = ' ';
+
+							argTypes[4] = CSTRINGOID;
+							argValues[4] = CStringGetDatum(contentType);
+							if(!contentType)
+								nulls[4] = 'n';
+							else
+								nulls[4] = ' ';
```

Comment:
The [SPI_execute_with_args](https://www.postgresql.org/docs/current/spi-spi-execute-with-args.html) function uses this interface for specifying nulls - `'n'` is a NULL and `' '`(whitespace char) is a non-null. Weird, but it is what it is.